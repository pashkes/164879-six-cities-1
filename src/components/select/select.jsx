import React from "react";
import PropTypes from "prop-types";

import withSelect from "./../../hocs/with-select/with-select";
import Option from "./../option/option.jsx";

export const Select = (props) => {
  const {
    onClickDropdown,
    isOpen,
    selected,
    onSelected,
    button,
    selectList,
    options,
    selectedName,
  } = props;

  return (
    <React.Fragment>
      <button type="button"
        className="places__sorting-type"
        onClick={onClickDropdown}
        aria-haspopup="true"
        aria-expanded={isOpen}
        id="sorter"
        ref={button}
      >
        {selectedName}
        <svg className="places__sorting-arrow" width="7" height="4" aria-hidden="true">
          <use xlinkHref="#icon-arrow-select"/>
        </svg>
      </button>
      <ul
        ref={selectList}
        aria-labelledby="sorter"
        className={`${isOpen ? `places__options--opened` : ``} places__options places__options--custom`}>
        {options.map((option, i) => {
          return <Option
            onSelected={onSelected}
            isSelected={selected === option.value}
            value={option.value}
            name={option.name}
            button={button}
            key={i}
          />;
        })}
      </ul>
      <select defaultValue={selected} hidden id="places-sorting">
        <option value={selected}>{selectedName}</option>
      </select>
    </React.Fragment>
  );
};

Select.propTypes = {
  onClickDropdown: PropTypes.func.isRequired,
  onSelected: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  selected: PropTypes.string.isRequired,
  button: PropTypes.object.isRequired,
  selectList: PropTypes.object.isRequired,
  selectedName: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string.isRequired,
    selected: PropTypes.bool.isRequired,
    name: PropTypes.string.isRequired,
  })),
};

export default withSelect(Select);
