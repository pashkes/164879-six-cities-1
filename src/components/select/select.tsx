import * as React from "react";

import withSelect from "./../../hocs/with-select/with-select";
import Option from "./../option/option";

interface Props {
  onOpenDropdown: () => void,
  onSelectOption: () => void,
  onKeyDownCloseDropdown: () => void,
  isOpen: boolean,
  selected: string,
  selectedName: string,
  options: {
    value: string,
    selected: boolean,
    name: string,
  }[] | null,
  onCloseDropdown: () => void,
}

export class Select extends React.PureComponent<Props, null> {
  readonly _button: React.RefObject<HTMLButtonElement>;
  readonly _dropdown: React.RefObject<HTMLUListElement>;

  constructor(props) {
    super(props);

    this._dropdown = React.createRef();
    this._button = React.createRef();
  }

  componentDidMount() {
    window.addEventListener(`click`, this.handleClickDocumentOutside);
  }

  componentWillUnmount() {
    window.removeEventListener(`click`, this.handleClickDocumentOutside);
  }

  handleClickDocumentOutside = (evt) => {
    const dropdown = this._dropdown.current.contains(evt.target);
    const button = this._button.current.contains(evt.target);
    const {isOpen, onCloseDropdown} = this.props;

    if (isOpen && !dropdown && !button) {
      onCloseDropdown();
    }
  };

  onFocusButton = () => {
    this._button.current.focus();
  };

  render() {
    const {
      onOpenDropdown,
      isOpen,
      selected,
      onSelectOption,
      options,
      selectedName,
      onKeyDownCloseDropdown,
    } = this.props;

    return (
      <React.Fragment>
        <button
          type="button"
          className="places__sorting-type"
          onClick={onOpenDropdown}
          aria-haspopup="true"
          aria-expanded={isOpen}
          id="sorter"
          ref={this._button}
        >
          {selectedName}
          <svg className="places__sorting-arrow" width="7" height="4" aria-hidden="true">
            <use xlinkHref="#icon-arrow-select"/>
          </svg>
        </button>
        <ul
          onKeyDown={onKeyDownCloseDropdown}
          ref={this._dropdown}
          aria-labelledby="sorter"
          className={`${isOpen ? `places__options--opened` : ``} places__options places__options--custom`}
          tabIndex={-1}
        >
          {options.map((option) => {
            return <Option
              onSelectOption={onSelectOption}
              isSelected={selected === option.value}
              onFocusButton={this.onFocusButton}
              value={option.value}
              name={option.name}
              button={this._button}
              key={option.value}
            />;
          })}
        </ul>
        <select defaultValue={selected} hidden id="places-sorting">
          <option value={selected}>{selectedName}</option>
        </select>
      </React.Fragment>
    );
  }
}

export default withSelect(Select);
