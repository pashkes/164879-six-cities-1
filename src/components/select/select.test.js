import React from "react";
import {shallow, configure} from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import {Select} from './select.jsx';
configure({adapter: new Adapter()});

it(`Select renders correctly`, () => {
  const option = shallow(<Select
    onSelected={jest.fn()}
    selected={``}
    onOpenDropdown={jest.fn()}
    onSelectOption={jest.fn()}
    selectedName={``}
    isOpen={false}
    onCloseDropdown={jest.fn()}
    options={[]}
    onKeyDownCloseDropdown={jest.fn()}
  />);
  expect(option).toMatchSnapshot();
});
