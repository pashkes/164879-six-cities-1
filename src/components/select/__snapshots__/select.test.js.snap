// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders correctly 1`] = `
<Fragment>
  <button
    aria-expanded={false}
    aria-haspopup="true"
    className="places__sorting-type"
    id="sorter"
    onClick={[MockFunction]}
    type="button"
  >
    <svg
      aria-hidden="true"
      className="places__sorting-arrow"
      height="4"
      width="7"
    >
      <use
        xlinkHref="#icon-arrow-select"
      />
    </svg>
  </button>
  <ul
    aria-labelledby="sorter"
    className=" places__options places__options--custom"
    onKeyDown={[MockFunction]}
    tabIndex="-1"
  />
  <select
    defaultValue=""
    hidden={true}
    id="places-sorting"
  >
    <option
      value=""
    />
  </select>
</Fragment>
`;
